src/pip/_internal/resolution/resolvelib/provider.py

```diff
@@ -234,8 +235,10 @@ def _eligible_for_upgrade(identifier: str) -> bool:
             constraint=constraint,
             prefers_installed=(not _eligible_for_upgrade(identifier)),
             incompatibilities=incompatibilities,
+            is_satisfied_by=self.is_satisfied_by,
         )
 
+    @lru_cache(maxsize=None)
```

Comment:
It would be good to check how much this increases memory consumption for a worst case scenario, i.e. when `ResolutionTooDeep` is reached. 

This set of requirements can reach that ```sphinx sphinx-rtd-theme sphinx-toolbox myst-parser sphinxcontrib-bibtex nbsphinx```. I can test this in the new year if you'd like.

Reply:
I experimented with a few different requirements dry install, looking at `apache-airflow[all]` I saw an increase of peak memory usage from 298 MBs to 306 MBs, I also saw the time to completion drom from ~2m 41s to ~2m 12s. 

Personally this seems significantly worth the memory/time trade off to me (and looking at the memray flamegraph of these installs there are potentially big areas for improvement in other parts of Pip's codebase to use less memory)